crates/threads-xform/src/lib.rs

```diff
@@ -0,0 +1,675 @@
+        instrs.push(Instruction::I32Const((stack_size / PAGE_SIZE) as i32));
+        instrs.push(Instruction::GrowMemory(0));
+        instrs.push(Instruction::SetLocal(0));
+
+        // if local0 == -1 then trap
+        instrs.push(Instruction::Block(BlockType::NoResult));
+        instrs.push(Instruction::GetLocal(0));
+        instrs.push(Instruction::I32Const(-1));
+        instrs.push(Instruction::I32Ne);
+        instrs.push(Instruction::BrIf(0));
+        instrs.push(Instruction::Unreachable);
+        instrs.push(Instruction::End); // end block
+
+        // stack_pointer = local0 + stack_size
+        instrs.push(Instruction::GetLocal(0));
+        instrs.push(Instruction::I32Const(PAGE_SIZE as i32));
+        instrs.push(Instruction::I32Mul);
+        instrs.push(Instruction::I32Const(stack_size as i32));
+        instrs.push(Instruction::I32Add);
+        instrs.push(Instruction::SetGlobal(stack_pointer_idx));
+    }
+
+    instrs.push(Instruction::Else);
+    for segment in segments {
+        // offset into memory
+        instrs.push(Instruction::I32Const(segment.offset as i32));
+        // offset into segment
+        instrs.push(Instruction::I32Const(0)); // offset into segment
+        // amount to copy
+        instrs.push(Instruction::I32Const(segment.len as i32));
+        instrs.push(Instruction::MemoryInit(segment.idx));
+    }
+    instrs.push(Instruction::End); // endif
+
+    // On all threads now memory segments are no longer needed
+    for segment in segments {
+        instrs.push(Instruction::MemoryDrop(segment.idx));
+    }
+
+    // If a start function previously existed we're done with our own
+    // initialization so delegate to them now.
+    if let Some(idx) = module.start_section() {
+        instrs.push(Instruction::Call(idx));
+    }
+
+    // End the function
+    instrs.push(Instruction::End);
+
+    // Add this newly generated function to the code section ...
+    let instrs = Instructions::new(instrs);
+    let local = Local::new(1, ValueType::I32);
+    let body = FuncBody::new(vec![local], instrs);
+    let code_idx = {
+        let s = module.code_section_mut().expect("module had no code");
+        s.bodies_mut().push(body);
+        (s.bodies().len() - 1) as u32 // TODO: checked cast
+    };
+    // ... and also be sure to add its signature to the function section ...
+    let type_idx = {
+        let section = module.type_section_mut().expect("module has no type section");
+        let pos = section.types()
+            .iter()
+            .map(|t| {
+                match t {
+                    Type::Function(t) => t,
+                }
+            })
+            .position(|t| t.params().is_empty() && t.return_type().is_none());
+        match pos {
+            Some(i) => i as u32,
+            None => {
+                let f = FunctionType::new(Vec::new(), None);
+                section.types_mut().push(Type::Function(f));
+                (section.types().len() - 1) as u32
```

Comment:
```suggestion
                (section.types().len() - 1) as u32 // TODO: checked cast
```

Reply:
Eh I've opted to just leave out the TODO for now, the chance of this actually getting hit is basically zero